### ISSUE TITLE
Sensorgnome RPI3 Receiver GPS Rollover Event Resultings in Incorrect Dates

### ISSUE THREAD

**Issue by leberrigan** [MotusDev/Motus-TO-DO#533]
See #516. 

[Data uploaded from receiver SG-91E6RPI317D2](https://sgdata.motus.org/status?jobID=973425) to _project 19 - Dossman Jamaica_ was affected by GPS rollover event, resulting in all data after **12:00 AM April 7th** appearing to come from **10:00 AM August 23, 1999**, onwards.

Is there a way we can modify the dates for these data? There should be a bunch of hits from after April 7th that isn't available right now because of this issue.

This has likely gone unnoticed with other receivers as well - perhaps we should do a database-wide check.

**Comment by denislepage** [comment]
This is not simple, but it’s likely doable.

We should have a database script that handles all the necessary steps to apply the correction, including modifying all the related tables that are impacted. That may also require changing the records in the Maria DB database. We should likely only change records on a batch basis, to make it easier to track what has been changed so far (e.g. have a table that track which batches have been corrected so far).

The ideal correction would be to handle this in the tag finder script for new incoming data.

**Comment by denislepage** [comment]
Digging a bit more.

There are 4 batches processed after April for that receiver, but all of their timestamps are around the Linux time 0 (1970-01-01) and there are no detections.

Actually, looking at all batches, I don’t think tag finder is able to process these data as if they were from 1999. I have no batches initiated in 1999, and almost none between 1990 and 2008. That makes it a lot harder to track the rollover issue because we often get batches that are coming without a proper GPS fix. There are also no GPS fixes in the GPS table past April 6th, which suggests that out of range values are being filtered out.

Looking at the original data files, however, it’s easy to see that we have data going from 1999-08-23 to 1999-10-26. There’s also data in 2000-01-01, but these may well be something different (e.g. collected before the GPS had a fix).

So, I don’t think there’s any useful that we can do in the Motus database, because the data doesn’t reach there, and it’ll be hard for us to even identify that it is happening, other than batches of data will be coming in with messed up dates and no detections.

I’m punting this to Joey and Jeremy, so we can try to sort out a possible fix for those. It’s in many ways tied in to fixes that are also needed with the boot number issue, so it would make sense to address at the same time.

The correction, I think, implies adding 619,315,200 seconds (1024 weeks) when ts is between approximately 933465600 (1999-08-01) and some upper number. To be safe, that upper number could be current date – 10 years, so if we ever get a date that is more than 10 years back and greater than Aug 1999, we can apply the 1024 weeks correction.

Something like:

if ts > 933465600 and ts < (current_ts – 315576000) then ts += 619315200

There’s a Clock_Repair function in find_tags. That may be a place to look into adding this option, but Joey should be a bit more familiar with the code to know.

**Comment by joeybernard** [comment]
This is only a first pass at a possible solution.

For the lotek receivers, the initial intake of data happens in ltMergeFiles.R, where the field ts is set in the table DTAtags. Then, handleLtFindtags.R calls getParamOverrides.R, which calls sgTimeAtBoot.R. This file only pulls those database entries that are more recent than the date stored in MOTUS_SG_EPOCH (currently set at 1262304000 or 2010-01-01). So it looks like it will ignore anything that has been ingested from before January 1, 2010.

My thought is that we need to add some code in ltMergeFiles.R to catch these cases and reset the value being stored in the ts field to a more reasonable date. Does this make sense?

**Comment by denislepage** [comment]
Yes, I think so. One complication is that I think the code also tries to correct for data collected in absence of a proper GPS fix. Most of those periods are rather short, but I think Jeremy mentioned they can last for several weeks or months, so it makes sense to try to preserve those data hits. I don't know where this is handled relative to what you are describing. GPS rollover issues should have a distinct time range that makes them relatively easy to differentiate from a lack of GPS time

**Comment by leberrigan** [comment]
The researcher was hoping to analyze these data some time this month - is it easy to apply a quick fix to these data before you come up with a long-term solution?

**Comment by denislepage** [comment]
To be clear, the change would have to be apply in tag finder. Those records are filtered out of processing at the moment.

**Comment by joeybernard** [comment]
I see that we have two options, and both are relatively quick to implement. The first is to actually fix the date when it gets ingested by the processServer process and gets dumped into the job tables to be processed. The second option is to change the SQL query that currently filters out anything before January 1, 2010, and have it accept jobs that also fall within some range around the last rollover in 1999. The decision as to which one we should do I see as a policy decision as to whether it is OK to change incoming data, or whether we should leave it as is and simply work around it. Thoughts?

**Comment by denislepage** [comment]
While I am usually reluctant to modify the original data, I think this is a case where I would be comfortable making an exception. The problem with making the correction later on is that we’d have to multiply the places where this correction would need to be applied. I say we fix as close possible to the source, in this case.

**Comment by joeybernard** [comment]
I made a new branch in the motusServer repo, named gps_rollover. Since we still don't have a test site, I would suggest that someone else review the change. It is as minimal as I could make it.

Testing it in production could be done by checking out this branch and doing rpack/restart on processServer. This way, if it causes an issue, we can simply checkout master again and do rpack/restart. If it passes testing we can merge this branch into master and switch the production server over to the master branch.

**Comment by denislepage** [comment]
I figured it might impact the outer ts bounds assigned to the batch record, perhaps?

We can try with either options and see what happens.

We should track down a receiver batch that we know is impacted and give it a try.

**Comment by jsayers-bsc** [comment]
Out of curiosity, was a resolution to this issue found? If data with dates of August 23 1999 come are uploaded, does the correction occur automatically?

**Comment by denislepage** [comment]
Yes, it should be.
